### PR TITLE
Migrate Purchase task and dashboard utils to TS

### LINKED ISSUE #37190: Migrate `purchase.js` and `index.js` in ./tasks/fills to TS
### Describe the solution you'd like

Migrate `purchase.js`, and `index.js` in ./tasks/fills to TS

- [ ] [index.js](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/tasks/fills/index.js)
- [ ] [purchase.js](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-admin/client/tasks/fills/purchase.js)

## Acceptance Criteria

- [ ] *.js files are converted to *.ts or *.tsx
- [ ] Run `pnpm --filter=woocommerce/client/admin ts:check` without any errors
- [ ] Not break existing functionalities

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #37190.

This PR migrates the Purchase task to TS. To archive that, I migrated the `dashboard/utils.js` as well.

Besides, this PR also fixes a wrong type definition (`productTypes`) in `@woocommerce/data` onboarding store.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

CI should pass

1. Go to the WooCommerce onboarding wizard.
2. Continue to `Onboarding wizard > Product Types` tab (/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard&step=product-types)
3. Select a paid product type, such as `Memberships`
4. Complete OBW
5. Go to `WooCommerce > Home`
6. Observe that there is a `Add ... product to my store` task (for example: `Add Memberships and 1 more product to my store`) 
7. Click on that task.
8. Observe that a modal pop up with title `Would you like to add the following paid features to your store now?`
9. Click on `Buy now` button
10. It should redirect you to checkout page (https://woocommerce.com/checkout/...).

### CODE REVIEW

File: packages/js/data/src/onboarding/types.ts [hunk 1]
```diff
@@ -75,7 +75,7 @@ export type OnboardingState = {
 	profileItems: ProfileItems;
 	taskLists: Record< string, TaskListType >;
 	paymentMethods: Plugin[];
-	productTypes: OnboardingProductType[];
```
Comment (chihsuan) [inline packages/js/data/src/onboarding/types.ts:78]: Onboarding `productTypes` is an object of objects.

Screenshot 2023-04-14 at 16 59 01

Screenshot 2023-04-14 at 16 11 30

File: plugins/woocommerce-admin/client/dashboard/utils.ts [hunk 2]
```diff
@@ -151,6 +92,7 @@ export function getProductList(
 	let theme = null;
 	if ( onboarding && onboarding.themes ) {
 		theme = onboarding.themes.find(
+			// @ts-expect-error Ignore TS error here, as we're removing the theme step and these code from OBW in another PR.
```
Comment (chihsuan) [inline plugins/woocommerce-admin/client/dashboard/utils.ts]: See https://github.com/woocommerce/woocommerce/commit/e863d28347ef82187ce423fb9969fc462164752e

**Review by adrianduffell — APPROVED** [review]
This tested perfect, thanks @chihsuan! 🚀 

Nice catch with the `productTypes` definition

File: plugins/woocommerce-admin/client/dashboard/utils.ts [hunk 3]
```diff
@@ -108,54 +155,9 @@ export function getCategorizedOnboardingProducts(
...
-
-/**
- * Get the value of a price from a string, removing any non-numeric characters.
- *
- * @param {string} string Price string.
- * @return {number} Number value.
- */
-export function getPriceValue( string ) {
-	return Number( decodeEntities( string ).replace( /[^0-9.-]+/g, '' ) );
+	return {
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/dashboard/utils.ts:158]: Nice touch! This feels more react-like

File: plugins/woocommerce-admin/client/tasks/fills/tax/index.tsx [hunk 4]
```diff
@@ -164,11 +164,7 @@ const Tax: React.FC< TaxProps > = ( { onComplete, query, task } ) => {
 				component: WooCommerceTax,
 				isVisible:
 					! taxJarActivated && // WCS integration doesn't work with the official TaxJar plugin.
-					woocommerceTaxCountries.includes(
-						getCountryCode(
-							generalSettings?.woocommerce_default_country
-						)
-					),
+					woocommerceTaxCountries.includes( countryCode ),
```
Comment (ilyasfoo) [inline plugins/woocommerce-admin/client/tasks/fills/tax/index.tsx:167]: Good catch!

**Review by ilyasfoo — APPROVED** [review]
Nice work, LGTM!

**Pull request merged**